pandas-stubs/core/frame.pyi

```diff
@@ -1011,7 +1016,85 @@ class DataFrame(NDFrame, OpsMixin):
         squeeze: _bool = ...,
         observed: _bool = ...,
         dropna: _bool = ...,
-    ) -> _DataFrameGroupByScalar: ...
+    ) -> DataFrameGroupBy[Scalar]: ...
+    @overload
+    def groupby(  # type: ignore[misc]  # pyright: ignore[reportOverlappingOverload]
+        self,
+        by: DatetimeIndex,
+        axis: Axis = ...,
+        level: Level | None = ...,
+        as_index: _bool = ...,
+        sort: _bool = ...,
+        group_keys: _bool = ...,
+        squeeze: _bool = ...,
+        observed: _bool = ...,
+        dropna: _bool = ...,
+    ) -> DataFrameGroupBy[Timestamp]: ...
+    @overload
+    def groupby(  # type: ignore[misc]
+        self,
+        by: TimedeltaIndex,
+        axis: Axis = ...,
+        level: Level | None = ...,
+        as_index: _bool = ...,
+        sort: _bool = ...,
+        group_keys: _bool = ...,
+        squeeze: _bool = ...,
+        observed: _bool = ...,
+        dropna: _bool = ...,
+    ) -> DataFrameGroupBy[Timedelta]: ...
+    @overload
+    def groupby(  # type: ignore[misc]
+        self,
+        by: PeriodIndex,
+        axis: Axis = ...,
+        level: Level | None = ...,
+        as_index: _bool = ...,
+        sort: _bool = ...,
+        group_keys: _bool = ...,
+        squeeze: _bool = ...,
+        observed: _bool = ...,
+        dropna: _bool = ...,
+    ) -> DataFrameGroupBy[Period]: ...
+    @overload
+    def groupby(  # type: ignore[misc]
+        self,
+        by: IntervalIndex[IntervalT],
+        axis: Axis = ...,
+        level: Level | None = ...,
+        as_index: _bool = ...,
+        sort: _bool = ...,
+        group_keys: _bool = ...,
+        squeeze: _bool = ...,
+        observed: _bool = ...,
+        dropna: _bool = ...,
+    ) -> DataFrameGroupBy[IntervalT]: ...
+    @overload
+    def groupby(
+        self,
+        by: MultiIndex,
```

Comment:
Might as well just combine this with the last overload that returns `DataFrameGroupBy[tuple]`

Reply:
This will not work due to `MultiIndex` being a subclass of `Index` so then it would match the incorrect overload first.

Reply:
If we wanted to reorder the overloads to combine these two cases, we would have to stop using `GroupByObjectNonScalar`, since that also contains `Index`.

Reply:
Although it seems like `GroupByObjectNonScalar` is currently only used in `series.py` and `frame.py`, so maybe the answer is to just do the refactor and remove `Index` from this type union and add an overload to `Series.groupby` for the ambiguous case.

The logical next step would be to do exactly the same change for `SeriesGroupBy` anyways, so the type of the group identifier gets preserved if you access a column on `DataFrameGroupBy`, although making it a two parameter `Generic` might make it slightly more confusing to use, so I opted to leave that transformation for another PR in the future.

Reply:
Yes, I see what you're saying, and I'm fine to let it be.